drivers/crypto/crypto_rpi_pico_sha256.c

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright (c) 2025 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_sha256
+
+#include <zephyr/crypto/crypto.h>
+
+#include <hardware/sha256.h>
+#include <pico/bootrom/lock.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sha_rpi_pico_sha256, CONFIG_CRYPTO_LOG_LEVEL);
+
+/*
+ * This is based on the pico_sha256 implementation in the Pico-SDK.
+ */
+
+/* We add one 0x80 byte, then 8 bytes for the size */
+#define SHA256_PADDING_DATA_BYTES 9
+#define SHA256_BLOCK_SIZE_BYTES   64
+
+struct rpi_pico_sha256_data {
+	bool locked;
+	uint8_t cache_used;
+	union {
+		uint32_t word;
+		uint8_t bytes[4];
+	} cache;
+	size_t total_data_size;
+};
+
+static void write_to_hardware(struct rpi_pico_sha256_data *state, const uint8_t *data,
+			      size_t data_size_bytes)
+{
+	if (!state->cache_used && !(((uintptr_t)data) & 3u)) {
+		/* aligned writes */
+		while (data_size_bytes >= 4) {
+			/* write a whole word */
+			sha256_wait_ready_blocking();
+			sha256_put_word(*(const uint32_t *)data);
+			data += 4;
+			data_size_bytes -= 4;
+		}
+	}
+
+	while (data_size_bytes--) {
+		state->cache.bytes[state->cache_used++] = *data++;
+		if (state->cache_used == 4) {
+			state->cache_used = 0;
+			sha256_wait_ready_blocking();
+			sha256_put_word(state->cache.word);
+		}
+	}
+}
+
+static inline size_t bytes_to_write(struct rpi_pico_sha256_data *state, size_t data_size_bytes)
+{
+	return ROUND_UP(state->total_data_size, SHA256_BLOCK_SIZE_BYTES) - state->total_data_size;
+}
+
+static void update_internal(struct rpi_pico_sha256_data *state, const uint8_t *data,
+			    size_t data_size_bytes)
+{
+	/* must finish off the last 64 byte block first or else sha256_err_not_ready will be true */
+	const size_t bytes_left = MIN(bytes_to_write(state, data_size_bytes), data_size_bytes);
+
+	if (bytes_left > 0) {
+		write_to_hardware(state, data, bytes_left);
+		state->total_data_size += bytes_left;
+		data_size_bytes -= bytes_left;
+		data += bytes_left;
+	}
+
+	/* Write the rest of the data */
+	if (data_size_bytes > 0) {
+		write_to_hardware(state, data, data_size_bytes);
+		state->total_data_size += data_size_bytes;
+	}
+}
+
+static void add_zero_bytes(struct rpi_pico_sha256_data *state, size_t data_size_bytes)
+{
+	const uint8_t zero[] = {0, 0, 0, 0};
+
+	while ((int32_t)data_size_bytes > 0) {
+		update_internal(state, zero, MIN(4, data_size_bytes));
+		data_size_bytes -= 4;
+	}
+}
+
+static void write_padding(struct rpi_pico_sha256_data *state)
+{
+	/* Has to be a multiple of 64 bytes */
+
+	uint64_t size = ROUND_UP(state->total_data_size, SHA256_PADDING_DATA_BYTES);
```

Comment:
```suggestion
	/* Has to be a multiple of 64 bytes */
	uint64_t size = ROUND_UP(state->total_data_size, SHA256_PADDING_DATA_BYTES);
```

I also have a question: the comment says that `size` must be multiple of 64 bytes, but `SHA256_PADDING_DATA_BYTES` is 9. Shouldn't we use `SHA256_BLOCK_SIZE_BYTES` instead here?

Reply:
This was a mistake when updating. Thanks for pointing that out.